core/node/consensus/src/storage/vm_reader.rs

```diff
@@ -0,0 +1,294 @@
+    // TODO(moshababo|BFT-493): optimize this process to reuse a single execution sandbox
+    // across one or multiple read sessions.
+    pub async fn read_committees(
+        &self,
+        ctx: &Ctx,
+        block_id: BlockId,
+    ) -> anyhow::Result<(Vec<CommitteeValidator>, Vec<CommitteeAttester>)> {
+        let mut conn = self.pool.connection(ctx).await.wrap("connection()")?.0;
+        let start_info = BlockStartInfo::new(&mut conn, Duration::from_secs(10))
+            .await
+            .unwrap();
+        let block_args = BlockArgs::new(&mut conn, block_id, &start_info)
+            .await
+            .unwrap();
+
+        let validator_committee = self
+            .read_validator_committee(block_args)
+            .await
+            .context("read_validator_committee()")?;
+        let attester_committee = self
+            .read_attester_committee(block_args)
+            .await
+            .context("read_attester_committee()")?;
+
+        Ok((validator_committee, attester_committee))
+    }
+
+    async fn read_validator_committee(
+        &self,
+        block_args: BlockArgs,
+    ) -> anyhow::Result<Vec<CommitteeValidator>> {
+        let mut committee = vec![];
+        let validator_committee_size = self
+            .read_validator_committee_size(block_args)
+            .await
+            .context("read_validator_committee_size()")?;
+        for i in 0..validator_committee_size {
+            let committee_validator = self
+                .read_committee_validator(block_args, i)
+                .await
+                .context("read_committee_validator()")?;
+            committee.push(committee_validator)
+        }
+        Ok(committee)
+    }
+
+    async fn read_attester_committee(
+        &self,
+        block_args: BlockArgs,
+    ) -> anyhow::Result<Vec<CommitteeAttester>> {
+        let mut committee = vec![];
+        let attester_committee_size = self
+            .read_attester_committee_size(block_args)
+            .await
+            .context("read_attester_committee_size()")?;
+        for i in 0..attester_committee_size {
+            let committee_validator = self
+                .read_committee_attester(block_args, i)
+                .await
+                .context("read_committee_attester()")?;
+            committee.push(committee_validator)
+        }
+        Ok(committee)
+    }
+
+    async fn read_validator_committee_size(&self, block_args: BlockArgs) -> anyhow::Result<usize> {
+        let func = self
+            .registry_contract
+            .function("validatorCommitteeSize")
+            .unwrap()
+            .clone();
+
+        let tx = self.gen_l2_call_tx(self.registry_address, func.short_signature().to_vec());
+
+        let res = self.eth_call(block_args, tx).await;
+
+        let tokens = func.decode_output(&res).context("decode_output()")?;
+        U256::from_tokens(tokens)
+            .context("U256::from_tokens()")
+            .map(|t| t.as_usize())
```

Comment:
Would it be feasible to do a bit more encapsulation? For example to create a `DeployedContract` or similar type which has an `Address` and a `Contract`, then wrap that into a `RegistryContract` which has methods for these that take whatever is needed to do the `eth_call` but otherwise know which function to call and how to parse the result? Maybe we can use the `abigen!` macro to generate the bindings if the ABI JSON is available at compile time? 

The `VMReader` sounds more general than just having all these methods for this particular contract, even if it's the only one at the moment. 

I did something similar [here](https://github.com/aakoshh/ipc/blob/main/fendermint/vm/interpreter/src/fvm/state/ipc.rs) where a gateway contract implemented multiple facets, each with their own binding, each represented by a [contract caller](https://github.com/aakoshh/ipc/blob/f13c0b6ecdbe16a18e2f3bd7eb6e898cda50552a/fendermint/vm/interpreter/src/fvm/state/fevm.rs#L136) that took the equivalent of the `TxSender` as an input. I suppose that gateway is similar to the `VMReader`, but a lot of the boilerplate and the private methods here could be delegated to contract specific structs, leaving only the highest level stuff in the reader.